### ISSUE TITLE
First method of class not highlighted in C++

### ISSUE THREAD

**Issue by Anteru** [pygments/pygments-migration-test#791]
(Original issue 1084 created by None on 2015-02-04T16:01:28.463861+00:00)

In the following code the first method(A) is not highlighted using version 2.0.1
Reproduced as http://pygments.org/demo/1221664/

```
#!c++

struct Type {
private:
  void A(int i, double j);
  void B(int i, double j);
};
```

**Comment by Anteru** [comment]
(Original comment by tshatch on 2015-10-18T16:27:42.592181+00:00)

Issue #1104 was marked as a duplicate of this issue.

**Comment by Anteru** [comment]
(Original comment by tshatch on 2015-11-01T03:43:58.887492+00:00)

I've done some debugging, and this is because 'root' transitions to 'statement' at the start of the struct (because it doesn't match the function patterns) and stays in there until the first semicolon.  I can fix these examples, but it does change (for example) sort_2 in example.c (to highlight the function name when *calling* to be Name.Function).  I believe this to be because of the separate types in the signature.


```
#!diff

diff -r 73115a27a7dd pygments/lexers/c_cpp.py
--- a/pygments/lexers/c_cpp.py  Sat Oct 31 17:49:48 2015 -0700
+++ b/pygments/lexers/c_cpp.py  Sat Oct 31 20:37:09 2015 -0700
@@ -104,7 +104,9 @@
         'statement': [
             include('whitespace'),
             include('statements'),
-            ('[{}]', Punctuation),
+            # Workaround for structs, classes, etc.
+            (r'[{:]', Punctuation, '#pop'),
+            ('\}', Punctuation),
             (';', Punctuation, '#pop'),
         ],
         'function': [
```
and the signature it mishighlights after that is


```
#!c

static int
sort_2(ap, bp, data)
    VALUE *ap, *bp;
    struct ary_sort_data *data;
{
...
	return rb_str_cmp(a, b);
```
I suspect a fix (recognizing this other signature) will need to happen after the 2.1 cut, unless @birkenfeld sees something obvious.

**Comment by Anteru** [comment]
(Original comment by None on 2016-05-03T11:28:05.371142+00:00)

Any new ideas on this subject?
I'm not sure I understand what the regression for the sort_2 example is.
When trying pygments with the source of example.c a lot of the functions didn't have the names highlighted. Does Pygments still has to support  K&R style declarations? At least in C++ they are not valid?

**Comment by LDVSOFT** [comment]
Any progress here too?

**Comment by Anteru** [comment]
No, this bug is 3 years old, but I didn't have a change to look at it, and this probably rather involved. If it's an urgent issue for you, would you consider filing a PR? I'm happy to review it.

**Comment by LDVSOFT** [comment]
Uh, I would've happily done so, but I don't have enough spare time to dive into :/
